File: chapter04/python/FindErrorNums.py
"""
对于这种数组问题，关键点在于元素和索引是成对儿出现的，常用的方法是排序、异或、映射。
映射的思路就是我们刚才的分析，将每个索引和元素映射起来，通过正负号记录某个元素是否被映射。
"""


def find_error_eums(nums):
    n = len(nums)
    dup = -1
    for i in range(n):
        idx = abs(nums[i]) - 1
        if nums[idx] < 0:
            dup = abs(nums[i])
        else:
            nums[idx] *= -1

    missing = -1
    for i in range(n):
        if nums[i] > 0:
            missing = i + 1
    return [dup, missing]


if __name__ == "__main__":
    nums = [4, 1, 4, 2]
    print(find_error_eums(nums))

File: chapter04/cpp/findErrorNums.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include "limits.h"
#include <math.h>

using namespace std;

vector<int> findErrorNums(vector<int> nums)
{
    int dup = 0;
    int mis = 0;
    int n = nums.size();
    for(int i=0; i<n; i++)
    {
        int idx = abs(nums[i]);
        if(nums[idx] < 0)
            dup = abs(nums[i]);
        else
            nums[idx] *= -1;
    }

    for(int i=0; i<n; i++)
    {
        if(nums[i] > 0)
        {
            mis = i;
            break;
        }
    }
    return vector<int>({dup, mis});
}

int main()
{   
    int a[] = {0, 4, 1, 4, 2};
    vector<int> nums(5, 0);
    for(int i=0; i<5; i++)
        nums[i] = a[i];

    vector<int> res = findErrorNums(nums);
    cout << res[0] << endl;
    cout << res[1] << endl;
    return 0;
}